netket/drivers/abstract_variational_driver.py

```diff
@@ -0,0 +1,276 @@
+        write_every=50,  # for default logger
+        step_size=1,  # for default logger
+        callback=lambda *x: True,
+    ):
+        """
+        Executes the Monte Carlo Variational optimization, updating the weights of the network
+        stored in this driver for `n_iter` steps and dumping values of the observables `obs`
+        in the output `logger`. If no logger is specified, creates a json file at `out`,
+        overwriting files with the same prefix.
+
+        Args:
+            :n_iter: the total number of iterations
+            :out: A logger object, or an iterable of loggers, to be used to store simulation log and data.
+                If this argument is a string, it will be used as output prefix for the standard JSON logger.
+            :obs: An iterable containing all observables that should be computed
+            :save_params_every: Every how many steps the parameters of the network should be
+            serialized to disk (ignored if logger is provided)
+            :write_every: Every how many steps the json data should be flushed to disk (ignored if
+            logger is provided)
+            :step_size: Every how many steps should observables be logged to disk (default=1)
+            :show_progress: If true displays a progress bar (default=True)
+            :callback: Callable or list of callable callback functions to stop training given a condition
+        """
+
+        if not isinstance(n_iter, numbers.Number):
+            raise ValueError(
+                "n_iter, the first positional argument to `run`, must be a number!"
+            )
+
+        if obs is None:
+            obs = {}
+
+        if out is None:
+            out = tuple()
+            print(
+                "No output specified (out=[apath|nk.logging.JsonLogger(...)])."
+                "Running the optimization but not saving the output."
+            )
+
+        # Log only non-root nodes
+        if self._mynode == 0:
+            # if out is a path, create an overwriting Json Log for output
+            if isinstance(out, str):
+                loggers = (_JsonLog(out, "w", save_params_every, write_every),)
+            else:
+                loggers = _to_iterable(out)
+        else:
+            loggers = tuple()
+            show_progress = False
+
+        callbacks = _to_iterable(callback)
+        callback_stop = False
+
+        with tqdm(
+            self.iter(n_iter, step_size), total=n_iter, disable=not show_progress
+        ) as itr:
+            for step in itr:
+
+                log_data = self.estimate(obs)
+
+                # if the cost-function is defined then report it in the progress bar
+                if self._loss_stats is not None:
+                    itr.set_postfix_str(self._loss_name + "=" + str(self._loss_stats))
+                    log_data[self._loss_name] = self._loss_stats
+
+                for logger in loggers:
+                    logger(self.step_count, log_data, self.machine)
```

Comment:
This and the loggers still need to be updated (self.machine does not exist here anymore)